File: Introduction/12__ePowerX.hs

-- https://www.hackerrank.com/challenges/eval-ex/

solve :: Double -> Double
solve 0.0 = 1.0
solve x = round' (sum [x ^ y / fromIntegral (factorial y) | y <- [0..9]])

factorial :: Integral a => a -> a
factorial 0 = 1
factorial n = n * factorial (n - 1)

-- rounds the number to better match the responses
round' :: RealFloat a => a -> a
round' num = (fromIntegral (round (num * 10000))) / 10000

main :: IO ()
main = getContents >>= mapM_ print. map solve. map (read::String->Double). tail. words

File: Introduction/12__ePowerX.scala

// https://www.hackerrank.com/challenges/eval-ex/

import scala.annotation.tailrec
import math.pow

def factorial(num: Int): Int = {

    @tailrec
    def tailRecFact(innerNum: Int, acc: Int): Int = {
        innerNum match {
            case 0 => acc
            case _ => tailRecFact((innerNum-1), acc * innerNum)
        }
    }
    tailRecFact(num, 1)
}

def ePowerX(num: Double): Double = {

    def findPower(count: Double): Double = {
        pow(num, count) / factorial(count.toInt)
    }

    @tailrec
    def tailRecEPowerX(count: Double, acc: Double): Double = {
        count match {
            case 0 => acc
            case _ => tailRecEPowerX((count-1), (acc + findPower(count)))
        }
    }
    tailRecEPowerX(9, 0)
}
